### DISCUSSION TITLE
Feature Request (Lights) - Presets/DIY ( Save & Restore )

### DISCUSSION THREAD

**Discussion by Montreal666** [arendst/Tasmota#14556] [Ideas]
If not already on the roadmap or implemented, any chance to add "presets" or "DIY" as generally seen in stock/cloud based light controller interfaces  (Magichome, etc. ) or WLED ( see examples below )

This would essentially save light features values (HSB, Color, CT, etc) in a defined number of "presets" so it can be saved/restored.
with the usual topic/cmnd associated.

ex: 

lightpresetX save  would save current values (HSB, Color, CT, etc) to preset X

lightpresetX restore  would restore values (HSB, Color, CT, etc) from preset X

This would be very useful for the Tasmota GUI and home automation integration.

I guess this could affect the memory required, especially if added to the GUI, so I would suggest to put it as a setoptionX to enable/disable.

What do you think?

![Capture d’écran 2022-01-21 113621](https://user-images.githubusercontent.com/56667580/150565252-c47d2da4-a4a1-4be1-bc61-d5c0b74cc351.png)
![Capture d’écran 2022-01-21 112747](https://user-images.githubusercontent.com/56667580/150565255-86389f85-7a0a-474c-b9ea-87add1a4d182.png)
![Capture d’écran 2022-01-21 114753](https://user-images.githubusercontent.com/56667580/150566879-d8fb3af0-bb7b-409b-9a2c-20bfcc2f35cb.png)

**Comment by sfromis** [comment]
As the assumption for Tasmota is that daily operation is done via a backend system, you can "easily" create presets there, publishing appropriate commands to Tasmota. The very rudimentary web UI is basically meant for admin/test, not as a device dashboard.

(What you can select via SetOptionX already has to be in the binary, and would not save flash. That would take a build time `#define` instead).

**Comment by ascillato** [comment]
That is doable with RULES and MEMs.

Let's say Mems are the custom preset color values:

```js
Mem1 255,255,255
Mem2 255,0,0
Mem3 0,255,0
```

You can modify these Mems by console command, HTTP or MQTT (`cmnd/your_topic/MEM1 255,255,255`)

Then the rules to trigger this presets:

```js
Rule1 1
Rule1 ON Event#LightPreset=1 DO Color %mem1% ENDON ON Event#LightPreset=2 DO Color %mem2% ENDON ON Event#LightPreset=3 DO Color %mem3% ENDON
```

So, after that, you can call each preset by MQTT as `cmnd/your_topic/EVENT LightPreset=1`
( where the topic is `cmnd/your_topic/EVENT` and the payload is `LightPreset=1` )

-----

If what you want is to have also Buttons in the WebUI to do that, you should use scripting instead of rules.

**Reply by Montreal666** [reply]
> That is doable with RULES and MEMs.
> 
> Let's say Mems are the custom preset color values:
> 
> ```js
> Mem1 255,255,255
> Mem2 255,0,0
> Mem3 0,255,0
> ```
> 
> You can modify these Mems by console command, HTTP or MQTT (`cmnd/your_topic/MEM1 255,255,255`)
> 
> Then the rules to trigger this presets:
> 
> ```js
> Rule1 1
> Rule1 ON Event#LightPreset=1 DO Color %mem1% ENDON ON Event#LightPreset=2 DO Color %mem2% ENDON ON Event#LightPreset=3 DO Color %mem3% ENDON
> ```
> 
> So, after that, you can call each preset by MQTT as `cmnd/your_topic/EVENT LightPreset=1` ( where the topic is `cmnd/your_topic/EVENT` and the payload is `LightPreset=1` )
> 
> If what you want is to have also Buttons in the WebUI to do that, you should use scripting instead of rules.

Is there a way to assign the current "color" value to a MEM with a replacement variable ?

so for instance if the current color value is 255,0,255

then sending

mem1 %color% 

would make mem1  = 255,0,255

use case is essentially to "push" values based on current state and "pop" them later on demand.

Open to any other approach, thx

**Reply by ascillato** [reply]
`%color%` is not a variable in Tasmota (yet). Besides, variables are only supported inside a rule (not from a MQTT/HTTP or Console command). The full list of available rule's variables is https://tasmota.github.io/docs/Rules/#rule-variables

If you want, you can ask Tasmota the actual color and it will respond:

```haskell
20:22:01.910 CMD: color
20:22:01.922 MQT: stat/ledliving/RESULT = {"Color":"255,255,255"}

```
So, you can parse that in your backend and resend it to Tasmota. I know you want to streamline things, so this approach don't seems to helps.

May be, we just need to add `%color%` as another variable for rules.

If we add `%color%` as a new variable for rules, you can add other rules for setting the presets like:

```js
Rule2 1
Rule2 ON Event#SetLightPreset=1 DO Mem1 %color% ENDON ON Event#SetLightPreset=2 DO Mem2 %color% ENDON ON Event#SetLightPreset=3 DO Mem3 %color% ENDON
```

Let me add it. It is a very small change to add `%color%` to rules. Stay tuned.

**Comment by Montreal666** [comment]
Thanks for the feedback, 
already had a backend solution but was looking for ways to streamline things;
The MEM/RULE approach is a good pointer, thanks!

**Reply by sfromis** [reply]
By exploiting the color command JSON output, you can also do this:
`on Color#data do Mem1 %value% endon`
You do not need a detour via backend.

However, this assumes that you have a JSON payload with only a single key, `Color`, meaning that a Color command was executed. It will not work when the Color field is part of a larger payload, where you instead need a rule like this:
`on Color do Mem1 %value% endon`

**Comment by ascillato** [comment]
PR Done: https://github.com/arendst/Tasmota/pull/14572

Wait to that PR be merged and try with latest version from development branch ( http://ota.tasmota.com/tasmota/tasmota.bin.gz )

-----

As a summary:

Rules for setting and for triggering 3 presets:
```js
Rule1 1
Rule1 ON Event#LightPreset=1 DO Color %mem1% ENDON ON Event#LightPreset=2 DO Color %mem2% ENDON ON Event#LightPreset=3 DO Color %mem3% ENDON
Rule2 1
Rule2 ON Event#SetLightPreset=1 DO Mem1 %color% ENDON ON Event#SetLightPreset=2 DO Mem2 %color% ENDON ON Event#SetLightPreset=3 DO Mem3 %color% ENDON
```

To SET each preset with the actual COLOR, by MQTT:
* Topic is `cmnd/your_topic/EVENT` and the payload is `SetLightPreset=1` )

To TRIGGER each preset by MQTT:
* Topic is `cmnd/your_topic/EVENT` and the payload is `LightPreset=1` )

Preset will be stored in MEM1.

Hope this helps.

**Comment by Montreal666** [comment]
Thank you for this great addition, much appreciated.
Glad you saw value in implementing it.
Cheers,
